src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1365 @@
+                    trace!(
+                        "Block arg {} is going to be in {}",
+                        vreg,
+                        Allocation::stack(self.vreg_spillslots[vreg.vreg()])
+                    );
+                }
+                let temp_slot = self.temp_spillslots[vreg.class()][next_temp_idx[vreg.class()]];
+                let temp = Allocation::stack(temp_slot);
+                next_temp_idx[vreg.class()] += 1;
+                let vreg_spill = Allocation::stack(self.vreg_spillslots[vreg.vreg()]);
+                trace!(
+                    "{} which is going to be in {} inserting move to {}",
+                    vreg,
+                    vreg_spill,
+                    temp
+                );
+
+                self.edits
+                    .add_move(inst, vreg_spill, temp, vreg.class(), InstPosition::Before);
+                // All branch arguments should be in their spillslots at the end of the function.
+                if self.vreg_allocs[vreg.vreg()].is_none() {
+                    self.live_vregs.insert(*vreg);
+                    let slot = self.vreg_spillslots[vreg.vreg()];
+                    self.vreg_allocs[vreg.vreg()] = Allocation::stack(slot);
+                    self.vreg_to_live_inst_range[vreg.vreg()].1 = ProgPoint::before(inst);
+                } else if self.vreg_allocs[vreg.vreg()] != vreg_spill {
+                    self.edits.add_move(
+                        inst,
+                        self.vreg_allocs[vreg.vreg()],
+                        vreg_spill,
+                        vreg.class(),
+                        InstPosition::Before,
+                    );
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn alloc_inst(&mut self, block: Block, inst: Inst) -> Result<(), RegAllocError> {
+        trace!("Allocating instruction {:?}", inst);
+        let operands = Operands::new(self.func.inst_operands(inst));
+        let clobbers = self.func.inst_clobbers(inst);
+
+        for (op_idx, op) in operands.reuse() {
+            trace!("Initializing reused_input_to_reuse_op for {op}");
+            let OperandConstraint::Reuse(reused_idx) = op.constraint() else {
+                unreachable!()
+            };
+            self.reused_input_to_reuse_op[reused_idx] = op_idx;
+        }
+        for (op_idx, op) in operands.fixed() {
+            let OperandConstraint::FixedReg(preg) = op.constraint() else {
+                unreachable!();
+            };
+            self.reserve_reg_for_fixed_operand(op, op_idx, preg)?;
+            if self.allocatable_regs.contains(preg) {
+                self.lrus[preg.class()].poke(preg);
+            }
+        }
+        for (_, op) in operands.fixed() {
+            let OperandConstraint::FixedReg(preg) = op.constraint() else {
+                unreachable!();
+            };
+            // Eviction has to be done separately to avoid using a fixed register
+            // as a scratch register.
```

Comment:
> Shouldn't remove_clobbers_from_available_pregs come before this? Otherwise a clobbered register might be used as a scratch register.


I don't think there is any problem with a clobbered register being used as a scratch register.

> In fact I think eviction for fixed operands should happen in process_operand_allocation. The same logic is needed no matter if the eviction is due to a fixed register or if the allocator selected a register with an existing vreg on it.

Eviction for fixed registers have to be done first because when iterating over the use and def operands are iterated over in any order. An example of why:

```
1. def v1 (reg), def v0 (fixed: p0)
2. use v1 (fixed: p0)
```

In the above example, v1 is already in p0 at inst 1. If we process v1 in inst 1 first, then v1 will be assigned to p0, even though it's not in the `available_pregs` set. The reason for this is the way `allocd_within_constraint` is implemented. In `allocd_within_constraint`, we can't just return false if the preg isn't in `available_pregs` because of the possibility of multiple operands using the same vreg. For example:

```
use v1 (fixed: p0), v1 (reg), v1 (reg)
```

The first encounter of v1 will remove p0 from `available_preg`. But for the remaining two, it will be wasteful to allocate new registers for them, so a check is made: `self.vreg_in_preg[preg.index()] == op.vreg()` to decide whether or not to conclude that it is allowed to be there. This introduces a new invariant: before `process_operand_allocation` is called, a vreg will be in a preg only if it's not reserved for some other vreg in the instruction. This necessitates the removal of all pregs from fixed registers before the use and def iterations.

Reply:
> I don't think there is any problem with a clobbered register being used as a scratch register.

You are correct, my bad.

> In the above example, v1 is already in p0 at inst 1. If we process v1 in inst 1 first, then v1 will be assigned to p0, even though it's not in the available_pregs set. The reason for this is the way allocd_within_constraint is implemented. In allocd_within_constraint, we can't just return false if the preg isn't in available_pregs because of the possibility of multiple operands using the same vreg. For example:

What I had in mind for such cases is to first check if there is already a use of this vreg in the same instruction and just reuse that allocation directly, skipping all the other checks. But your method also works, so it's fine to leave it as it is.